pandas-stubs/core/base.pyi

```diff
@@ -19,13 +24,21 @@ from pandas._typing import (
     Scalar,
     npt,
 )
+from pandas.util._decorators import cache_readonly
+
+class PandasObject: ...
 
 class NoNewAttributesMixin:
-    def __setattr__(self, key, value) -> None: ...
+    def __setattr__(self, key: str, value: Any) -> None: ...
 
 class SelectionMixin(Generic[NDFrameT]):
+    obj: NDFrameT
+    exclusions: frozenset[Hashable]
+    @final
+    @cache_readonly
```

Comment:
I would be inclined to keep final and cache_readonly:

- `final`: could help downstream developers who extend pandas classes to know not to overwrite a method
- `cache_readonly`: we have to use at least `property` to indicate the correct type

Reply:
What @twoertwein said.
These are not added only to make stubtest happy.

`final` is actually a more strict interface. It tells the users that they are not supposed to override a function or a class otherwise they are on their own if something breaks.

`cache_readonly` is currently an alias to `property` so removing it has the same effect as removing a `property` decorator. I think using `cache_readonly` has two more advantages over using `property` apart from consistency with runtime. 1) its name is descriptive; it cahces the result and it is not writable, and 2) if/when type checkers start to recognize custom property definitions, this can be changed in one place instead of having to go through all the code base to change `property` to `cache_readonly` (both are actually used by pandas).

Reply:
OK, thanks for the explanation